src/main/kotlin/com/example/delivery/config/LogInvokeAnnotationBeanPostProccessor.kt

```diff
@@ -0,0 +1,88 @@
+package com.example.delivery.config
+
+import com.example.delivery.annotaion.InvokeLog
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.cglib.proxy.Enhancer
+import org.springframework.cglib.proxy.MethodInterceptor
+import org.springframework.core.Ordered
+import org.springframework.core.annotation.Order
+import org.springframework.stereotype.Component
+import java.lang.reflect.Method
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.memberProperties
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.isAccessible
+import kotlin.reflect.jvm.jvmErasure
+
+@Component
+@Order(Ordered.HIGHEST_PRECEDENCE)
+class LogInvokeAnnotationBeanPostProcessor : BeanPostProcessor {
+
+    private val beans = HashMap<String, AnnotationTarget>()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = bean.javaClass
+        when {
+            beanClass.isAnnotationPresent(InvokeLog::class.java) -> {
+                beans[beanName] = AnnotationTarget.CLASS
+            }
+
+            beanClass.declaredMethods.any { it.isAnnotationPresent(InvokeLog::class.java) } -> {
+                beans[beanName] = AnnotationTarget.FUNCTION
+            }
+        }
+        return bean
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any {
+        return if (beans.containsKey(beanName)) {
+            createProxy(bean, beanName)
+        } else {
+            bean
+        }
+    }
+
+    private fun createProxy(bean: Any, beanName: String): Any {
+        bean::class.memberProperties.forEach {
+            it.isAccessible = true
+        }
+
+        val constructorParams: List<KParameter> = bean::class.primaryConstructor!!.parameters
+        val argumentTypes: Array<Class<*>> = constructorParams.map { it.type.jvmErasure.java }.toTypedArray()
+        val arguments: Array<Any> = constructorParams.map { extractValue(it, bean) }.toTypedArray()
+
+        val enhancer = Enhancer()
+        enhancer.setSuperclass(bean.javaClass)
+        enhancer.setCallback(createInterceptor(beans.getValue(beanName)))
+
+        return enhancer.create(argumentTypes, arguments)
+    }
+
+    private fun extractValue(param: KParameter, bean: Any): Any {
+        return bean::class.memberProperties.find { it.name == param.name }!!.call(bean)!!
```

Comment:
The rule `#1`: 
if u use `!!` then it's high probability you are doing something wrong.

 ```suggestion
        return bean::class.memberProperties.first{ it.name == param.name }.call(bean)!!
```